### ISSUE TITLE
Lost / Dropped Video Frames (and fHDHR restart)

### ISSUE THREAD

**Issue by arrmo** [fHDHR/fHDHR#297]
Hi,

OK, this was sort of tied before to "lagging" video - but trying to make it a bit more concrete. Ya, some of this will be a bit fuzzy yet 😜. But we'll figure it out.

The issue that I am seeing is video degrading vs. time (lagging, blocky, etc.), and trying to quantify this using VLC statistics. I (think) I see a tie in to log level as well, but that's perhaps just due to overall load (of Python?). But that all said, some data to try to explain, and then some observations below that. Yell of course if I'm just off base. Overall things are working quite well now, just minor little items - except for the dropping video.

And the data. I tried to capture (from VLC), Lost Video packets, and Dropped Input/Read (as VLC calls them). I ran this with my Ceton PCIe card, and tried both direct and ffmpeg. Also a couple of log levels. And as a baseline, I always ran in parallel with fHDHR a test with direct stream (direct, from /dev to vlc), `clear; cat /dev/ceton/ctn91xx_mpeg0_3 | vlc -`. So now, the data,
![image](https://user-images.githubusercontent.com/8756384/156472776-d419ae7f-fcd2-405b-b0f3-9fdcd9039d22.png)

And, some observations - qualitative, and I maybe just just crazy, but hoping this helps,
1. The errors do seem to arrive in a bursty nature ... not dribbling in, but in "jumps"
2. The errors seem to "accelerate" vs. time ... starting slow, then getting worse. A good example is the last row - it ran quite clean for a long time ... then in the last ~ 40k frames it lost 3000 of them ... and actually, fHDHR restarted! I noticed it because the VLC video stopped, when to look at the log, and saw it had restarted. I have attached the full debug log for this one (sorry, I know it's long, but wanted to share it all in case it helps).
3. When the error count gets high, if I just stop the stream, start over ... it seems to be clean again.

I may be out to lunch here, but I'm wondering about memory leaks (typical for vs. time issues), but also ... perhaps run the generator in a separate thread? Or, use a Python queue? Then again, could it be because we aren't reading an integer number of packets each time? Again, just thinking out loud.

Unfortunately this tends to take a while to show up ... can be a few hours, so not quick to test. That said, let me know if there are things you want me to try.

Thanks!
[crash-log.txt](https://github.com/fHDHR/fHDHR/files/8173872/crash-log.txt)

**Comment by deathbybandaid** [comment]
I assume you've only tested with ceton? My tvheadend origin works without issue. I've also had fantastically long m3u8 streams.

The ffmpeg and vlc plugins using their respective third party app aren't reading that integer, but reading and serving that integer chunk back to fHDHR, and subprocess runs those applications in a seperate thread.

**Comment by deathbybandaid** [comment]
Have you tried the vlc plugin?

**Comment by arrmo** [comment]
> I assume you've only tested with ceton? My tvheadend origin works without issue. I've also had fantastically long m3u8 streams.

Correct! I ran the baseline to make sure the data source (HW, driver) are good - which they seem to be. Hmmm ... if it's tied to the ceton origin - is that code really in the "loop" as it streams?



> Have you tried the vlc plugin?

I have not, but can if that would help. Also ... could the chunks be fed to a file? Let it run for a while, then play back the file in VLC => that would help to identify the issue ... read or write. Agreed?

Thanks!

**Comment by deathbybandaid** [comment]
Writing that many chunks to disk would introduce extra I/O and kill performane as well as wear and tear on a disk

**Comment by arrmo** [comment]
> Writing that many chunks to disk would introduce extra I/O and kill performane as well as wear and tear on a disk

Agreed! I was just thinking as a debug exercise, not normally for sure. Just to capture what has been read, check it before write that way. Make sense?

**Comment by arrmo** [comment]
A thought about this. tvheadend, what is your bitrate? In the cases I am running, the ceton is sending in data at ~ 16-18 Mbps, and this is also the output rate (ffmpeg copy). Curious if the high bit rate is part of the issue => lots of data to process.

I did run something overnight last night. The baseline as above (/dev direct to VLC), and I saw lost/dropped of 152/1 in 2.7M packets. Pretty low, but I also ran /dev to ffmpeg, transcode (h264_nvenc) to ~ 2 Mbps, there only 13/0 in ~ 1.9M packets ... so ~ 10x lower. Just by changing the output rate.

Wondering if I should try the same in fHDHR => transcode down to a lower output rate, see if that is part of the trigger. Thoughts?

Thanks!

**Comment by deathbybandaid** [comment]
as of right now, my focus will be on tearing apart and restructuring core, testing for "stability" and publishing the next release.

once that is done, I can take a deeper look at this issue.

**Comment by arrmo** [comment]
That sounds like a lot of work (and pain)?

No worries, will dig here - see if I can get to the bottom of it. Was just wondering if this check made sense 😆.

Thanks!

**Comment by arrmo** [comment]
Hi,

OK, trying different cases, to see if I can work out if this is a read (input) or write (output) issue. So I'm running ffmpeg right now, forced (hacked in) HW encoding => to lower the output bit rate, see if that has an impact. Cautiously optimistic, but so far only 3 lost / 0 dropped frames, out of ~ 330k.

Of course this is with rtp (udp), as ffmpeg and rtp seem to be somewhat tied. I'd like to try direct next if I can, also going through ffmpeg => to run a similar test, but with direct. But I'm a bit perplexed, as it seems like ffmpeg is "tied" to rtp. Or am I misunderstanding?

Thanks!

**Comment by arrmo** [comment]
OK, a very interesting result - or at least I think so ... LOL. I tried as noted above, Ceton to ffmpeg (RTP, as required to trigger ffmpeg), and leaving transcode as "none" (copy), or I manually hacked in h264_nvenc (Nvidia h.264 HW encoding). I wanted to see if the ffmpeg output bit rate has an impact. It seems to 😉,
![image](https://user-images.githubusercontent.com/8756384/156821628-dfc4359c-ded0-49b8-aa4b-4f11faf90b78.png)

The cases here
1. hw encoding, h.264, as noted above. Left it over night. Very low drop out - looks great. Output bit rate is ~ 2.5-3 Mbps.
2. change back to copy "transcoding" ... nothing else at all changed. Quite a delta. And the errors seem to come in bursts, as noticed before. Output bit rate matches the source, ~ 16 Mbps or so.
3. back to 1 again, just a sanity check. Again. looks excellent.

So it does seem like the output bit rate has an impact. I need to dig a bit more, as it looks like fHDHR only sets up ffmpeg (if I read the command line right), and then ffmpeg pulls directly from RTP (correct?). So the chunks / buffering is only serving the output / client, is that right? Or do I have this wrong in my head? In any case, will dig further, but wanted to share this result - it is very encouraging.

Thanks!

**Comment by arrmo** [comment]
OK, pretty sure I found the issue here - have run (HW transcoded) video for ~ 4:05 now, 880k blocks ... 0 lost or dropped!

I started thinking about buffers, as that would be something that could cause a low rate of occurrence like this one. Some hunting around, and quite a few posts about issues with stdout buffering in subprocess, so I tried ...
```
ffmpeg_proc = subprocess.Popen(self.ffmpeg_command, stdout=subprocess.PIPE, bufsize=0)
```
Thinking I'd kill the buffer here, you already have one in your chunk serving. Fixed the issue with this! I still want to test with full rate transfer (i.e. "copy" transcoding), but it is looking quite good right now.

Thanks!

**Comment by arrmo** [comment]
Full data rate, copy "codec" in ffmpeg => 11 dropped out of 850k. So very low, looking quite good. Incorporate the update above (`bufsize=0`), then close this one out?

Thanks!
